src/coreclr/jit/codegencommon.cpp

```diff
@@ -4593,8 +4976,39 @@ void CodeGen::genEnregisterIncomingStackArgs()
         assert(regNum != REG_STK);
 
         var_types regType = varDsc->GetActualRegisterType();
+#ifdef TARGET_LOONGARCH64
+        {
+            bool FPbased;
+            // int baseOffset = (regArgTab[argNum].slot - 1) * slotSize;
+            int base = compiler->lvaFrameAddress(varNum, &FPbased);
 
+            if ((-2048 <= base) && (base < 2048))
+            {
+                GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, 0);
+            }
+            else
+            {
+                if (tmp_reg == REG_NA)
+                {
+                    regNumber reg2 = FPbased ? REG_FPBASE : REG_SPBASE;
+                    tmp_offset     = base;
+                    tmp_reg        = REG_R21;
+                    GetEmitter()->emitIns_R_I(INS_lu12i_w, EA_PTRSIZE, REG_R21, tmp_offset >> 12);
+                    GetEmitter()->emitIns_R_R_I(INS_ori, EA_PTRSIZE, REG_R21, REG_R21, tmp_offset & 0xfff);
+                    GetEmitter()->emitIns_R_R_R(INS_add_d, EA_PTRSIZE, REG_R21, REG_R21, reg2);
+                    GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, -8);
+                }
+                else
+                {
+                    int baseOffset = -(base - tmp_offset) - 8;
+                    GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, baseOffset);
+                }
+            }
+        }
+#else
```

Comment:
I've been trying to understand this code, and it concerns me, for a number of reasons:

1. tmp_reg is set to REG_R21, but the subsequent code uses REG_R21 directly: shouldn't it use tmp_reg?
2. It looks like you're setting up R21 as a temporary offset register for large offsets. But how does emitIns_R_S know to use it? Is it a "magic" parameter to emitIns_R_S? That is, emitIns_R_S expects it to be set sometimes?
3. Why is there a `-8` in `GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, -8);`? Shouldn't the `varNum` be used by emitIns_R_S to get the exact offset?
4. What if there are so many stack arguments that we set up R21 to load one, and then it's >2048 bytes away from R21 for the next one that we load into a register; won't there be a range limit failure?

Reply:
> I've been trying to understand this code, and it concerns me, for a number of reasons:
> 
> 1. tmp_reg is set to REG_R21, but the subsequent code uses REG_R21 directly: shouldn't it use tmp_reg?

R21 is reserved as a special scrapt register and using it should be very careful.

> 2. It looks like you're setting up R21 as a temporary offset register for large offsets. But how does emitIns_R_S know to use it? Is it a "magic" parameter to emitIns_R_S? That is, emitIns_R_S expects it to be set sometimes?

Yes, here the R21 is the "magic" register. and only for the large offset >2047 which can be encode within a load instruction.

> 3. Why is there a `-8` in `GetEmitter()->emitIns_R_S(ins_Load(regType), emitTypeSize(regType), regNum, varNum, -8);`? Shouldn't the `varNum` be used by emitIns_R_S to get the exact offset?

yes, the `varNum` is used by emitIns_R_S to get the exact offset.
When  the large offset >2047 which can be encode within a load instruction, if each offset should use another instructions to construct, there will waste the space and maybe eat up `SC_IG_BUFFER_SIZE`.
For LoongArch64, we don't want to extend the size of `SC_IG_BUFFER_SIZE` which waste space  while we optimize the instruction emiting especially for the repeating large offset constructing.

> 4. What if there are so many stack arguments that we set up R21 to load one, and then it's >2048 bytes away from R21 for the next one that we load into a register; won't there be a range limit failure?

Reply:
Don't worry about increasing `SC_IG_BUFFER_SIZE`: it's a single, global, re-used buffer. In fact, I just recently increased its size for arm/arm64: https://github.com/dotnet/runtime/pull/65153.

I guess the key to this code is you know there will likely be multiple large offsets very close together, so it is worthwhile to establish a "temporary frame pointer" in R21 and do multiple loads from that. Makes sense: it prevents creating lots of large constant constructions that are all very similar.

However, you need to make sure you re-establish R21 if you ever find an offset constant that is "too far" away from the one you already established.

How does emitIns_R_S know to use R21 instead of FP or SP? Instead of calling emitIns_R_S, couldn't you just use `emitIns_R_R_I(insLoad(), regNum, REG_R21, offset)`, which the appropriate offset given?

Reply:
First I will add some comment for LoongArch64.

> Don't worry about increasing `SC_IG_BUFFER_SIZE`: it's a single, global, re-used buffer. In fact, I just recently increased its size for arm/arm64: #65153.

Yes, this is a re-used buffer. But the instructions which emitted into the last executing space are redundant.
Maybe we can optimize the instructions about the large offset is the root cause.

> 
> I guess the key to this code is you know there will likely be multiple large offsets very close together, so it is worthwhile to establish a "temporary frame pointer" in R21 and do multiple loads from that. Makes sense: it prevents creating lots of large constant constructions that are all very similar.
> 
> However, you need to make sure you re-establish R21 if you ever find an offset constant that is "too far" away from the one you already established.

This is local case. Of course I have a another way planning for the global large offset to optimize further, it's just a under consideration. If need, maybe we can discuss it later seperately.

> 
> How does emitIns_R_S know to use R21 instead of FP or SP? Instead of calling emitIns_R_S, couldn't you just use `emitIns_R_R_I(insLoad(), regNum, REG_R21, offset)`, which the appropriate offset given?

The args `offs` of `emitIns_R_S` is the key factor for optimizing by R21.
The release offs is postive. If we negive it meaning using the R21 for optimize within the `emitIns_R_S`.
```
 853     base = emitComp->lvaFrameAddress(varx, &FPbased);
 854     imm  = offs < 0 ? -offs - 8 : base + offs;
 855 
 856     regNumber reg2 = FPbased ? REG_FPBASE : REG_SPBASE;
 857     reg2           = offs < 0 ? REG_R21 : reg2;
 858     offs           = offs < 0 ? -offs - 8 : offs;
 859 
```
All these are setted before calling the `emitIns_R_S`.

Reply:
I really don't like how this works:
1. I don't like the "magic" argument of R21 and negative offsets
2. The "-8" is non-obvious
3. As I mentioned, I think there are bugs in the code such that it won't handle all cases of large offsets.
4. It also doesn't handle the case of some locals being FP-relative and some being SP-relative (this is a special case in the prolog, so it probably doesn't matter)

How about a design like this:
1. Functions emitStartUsingExtraFramePointer() / emitStopUsingExtraFramePointer() -- call these at the beginning and end of genEnregisterIncomingStackArgs(). It would initialize data to allow you to use R21 as an "extra frame pointer".
2. Call emitIns_R_S() as usual.
3. Move all this logic into emitIns_R_S:
    a. If the offset from the base register fits (-2048 <= offset <= 2047), use it.
    b. if there already exists an "extra frame register", and the offset from the "extra frame register" fits, then use it. NOTE: the extra frame register might be established relative to SP, or relative to FP: it's important to make sure the offsets are compared properly.
    c. else, set (or set again) the "extra frame register" to the offset of the current local, and use it to load the local.